File: binarytree/0104-maxDepth.py
"""
leetcode 104
max Depth of a binary tree

why the recursive?

what is the base condition to stop the recursion?
the base condition forces the return from endless recursion.

1. Don't get into the details. Just think about the whole tree and 
its left and right subtrees.

maxDepth of the tree = max(maxDepth of left sub, maxDepth of right sub) + 1

2. 
Original problem: max depth of the tree
sub problem: max depth of either left or right sub trees

Both problems are similar. But the sub problem is embedded into another.
That is, one sub problem has to return its result to its upper-level.

The sub-problem becomes smaller and smaller, till it reaches the leaf.

For the recursive the computer is helping you and 
it keep tracking the return with the stack.
"""

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None: return 0
        l_depth = self.maxDepth(root.left)
        r_depth = self.maxDepth(root.right)
        return max(l_depth, r_depth) + 1

# 
# level traversal
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        # initialize 
        que = [root]
        depth = 0
        # start to traverse
        while que:
            n = len(que)
            for i in range(n):
                node = que.pop(0)
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
            depth += 1
        #
        return depth

# 
# another recursive
#
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        ans = 0 # global variable
        def f(node, cnt):
            if node is None:
                return
            cnt += 1
            nonlocal ans
            ans = max(ans, cnt)
            f(node.left, cnt)
            f(node.right, cnt)
        f(root, 0)
        return ans

File: binarytree/0104-maxDepth.cpp
/*
 * leetcode 104
 * 
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// concise, recursive
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) return 0;
        return max(maxDepth(root->left), maxDepth(root->right)) + 1;
    }
};

// expand to make it clear
class Solution {
public:
    int maxDepth(TreeNode *root) {
        if (root == nullptr) return 0;
        // after-order traversal
        int l_depth = maxDepth(root->left);
        int r_depth = maxDepth(root->right);
        return max(l_depth, r_depth) + 1;
    }
};

// template
class Solution {
public:
    // at one level of the recursive, just think about this level
    // just this level
    int depth(TreeNode* cur) {
        if (cur == nullptr) return 0;
        int l = depth(cur->left);
        int r = depth(cur->right);
        return max(l, r) + 1;
    }
    int maxDepth(TreeNode* root) {
        return depth(root);
    }
};
